jscomp/syntax/ast_utf8_string.ml

```diff
@@ -17,129 +17,230 @@
-  if offset + 1 >= s_len then 
+and two_hex loc buf s offset s_len =
+  if offset + 1 >= s_len then
     Location.raise_errorf ~loc "\\x need at least two chars";
-  let a, b = s.[offset], s.[offset + 1] in 
-  if Ext_char.valid_hex a && Ext_char.valid_hex b then 
-    begin 
-      Buffer.add_char buf a ; 
-      Buffer.add_char buf b ; 
-      check_and_transform loc buf s (offset + 2) s_len 
+  let a, b = s.[offset], s.[offset + 1] in
+  if Ext_char.valid_hex a && Ext_char.valid_hex b then
+    begin
+      Buffer.add_char buf a ;
+      Buffer.add_char buf b ;
+      check_and_transform loc buf s (offset + 2) s_len
     end
   else Location.raise_errorf ~loc "%c%c is not a valid hex code" a b
 
-and unicode loc buf s offset s_len = 
-  if offset + 3 >= s_len then 
+and unicode loc buf s offset s_len =
+  if offset + 3 >= s_len then
     Location.raise_errorf ~loc "\\u need at least four chars";
   let a0,a1,a2,a3 = s.[offset], s.[offset+1], s.[offset+2], s.[offset+3] in
-  if 
+  if
     Ext_char.valid_hex a0 &&
     Ext_char.valid_hex a1 &&
     Ext_char.valid_hex a2 &&
-    Ext_char.valid_hex a3 then 
-    begin 
+    Ext_char.valid_hex a3 then
+    begin
       Buffer.add_char buf a0;
       Buffer.add_char buf a1;
       Buffer.add_char buf a2;
-      Buffer.add_char buf a3;  
-      check_and_transform loc buf s  (offset + 4) s_len 
-    end 
-  else 
+      Buffer.add_char buf a3;
+      check_and_transform loc buf s  (offset + 4) s_len
+    end
+  else
     Location.raise_errorf ~loc "%c%c%c%c is not a valid unicode point"
-      a0 a1 a2 a3 
+      a0 a1 a2 a3
 (* http://www.2ality.com/2015/01/es6-strings.html
    console.log('\uD83D\uDE80'); (* ES6*)
    console.log('\u{1F680}');
-*)   
\ No newline at end of file
+*)
+
+type interpo = Text of string | Delim of string
+
+type ('a, 'b) either = Left of 'a | Right of 'b
+
+let consume_text s start_index =
+  let rec _consume_text s index last_char new_word =
+    if index = String.length s then new_word, String.length s
+    else begin
+      match s.[index] with
+      | '\\' -> (if index + 1 = String.length s then "", index else
+                   match s.[index+1] with
+                   | '\\' -> _consume_text s (index+2) ' ' (Ext_string.append_char new_word '\\')
+                   | '$' -> _consume_text s (index+2) ' ' (Ext_string.append_char new_word '$')
+                   | c -> _consume_text s (index+1) '\\' (Ext_string.append_char new_word '\\'))
+      | '$' -> (new_word, index)
+      | c -> _consume_text s (index + 1) c (Ext_string.append_char new_word c)
+    end
+  in _consume_text s start_index ' ' ""
+
+let consume_delim s start_index =
+  let with_par = ref false in
+  let rec _consume_delim s index ident =
+    if index = String.length s then (if !with_par = true then (None, index) else (Some ident, index))
+    else
+      match s.[index] with
+      | '(' -> (if !with_par = false then (with_par := true; _consume_delim s (index+1) ident) else (None, index))
```

Comment:
I am unclear about such piece of code, our interpolation is quite simple(no nested interpolation).
it is only `$x` or `$(x)`